File: dcp-solutions/2024-12-03/solution.js
```javascript
function numIslands(matrix) {
    if (!matrix || !matrix[0]) {
        return 0;
    }

    const rows = matrix.length;
    const cols = matrix[0].length;

    function dfs(r, c) {
        if (r < 0 || r >= rows || c < 0 || c >= cols || matrix[r][c] !== '1') {
            return;
        }

        matrix[r][c] = '0';

        dfs(r - 1, c);
        dfs(r + 1, c);
        dfs(r, c - 1);
        dfs(r, c + 1);
    }

    let islands = 0;

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (matrix[r][c] === '1') {
                dfs(r, c);
                islands++;
            }
        }
    }

    return islands;
}
```

File: dcp-solutions/2024-12-03/solution.c
```c
#include <stdio.h>

int numIslands(char** grid, int gridSize, int* gridColSize) {
    if (!grid || !grid[0]) return 0;
    int islands = 0;
    for (int i = 0; i < gridSize; ++i) {
        for (int j = 0; j < *gridColSize; ++j) {
            if (grid[i][j] == '1') {
                islands++;
                dfs(grid, i, j);
            }
        }
    }
    return islands;
}

void dfs(char** grid, int row, int col) {
    if (row >= 0 && row < *gridSize && col >= 0 && col < *gridColSize && grid[row][col] == '1') {
        grid[row][col] = '0';
        dfs(grid, row + 1, col);
        dfs(grid, row - 1, col);
        dfs(grid, row, col + 1);
        dfs(grid, row, col - 1);
    }
}
```